src/mon/MonCommands.h

```diff
@@ -1251,6 +1251,36 @@ COMMAND_WITH_FLAG("smart name=devid,type=CephString,req=false",
 COMMAND_WITH_FLAG("mon_status", "report status of monitors", "mon", "r",
 		  FLAG(NOFORWARD)|FLAG(HIDDEN))
 COMMAND_WITH_FLAG("heap "						\
-            "name=heapcmd,type=CephChoices,strings=dump|start_profiler|stop_profiler|release|stats", \
```

Comment:
I don't understand the purpose of this patch. If we add the commands here, they are included in the 'ceph -h' CLI help.. but we don't want them there.  Their signatures are passed register_command and the asok machinery checks them.

The asok declarations don't include r/rw/rwx info because they are all assumed to require ``*``... that's partly why the permissions check ended up the way it did (the cap check was an afterthought so that we could allow the ``smart`` tell command via a cap).

Reply:
> I don't understand the purpose of this patch. If we add the commands here, they are included in the 'ceph -h' CLI help..

no, they are all hidden. as in `ceph.in`, all commands marked with `OBSOLETE`,  `DEPRECATED` or `HIDDEN` are not listed in `--help` output.

> The asok declarations don't include r/rw/rwx info because they are all assumed to require ``*``... that's partly why the permissions check ended up the way it did (the cap check was an afterthought so that we could allow the ``smart`` tell command via a cap).

@liewegas the reason is to unify the checks (including caps check and deprecation checks). i think it'd be better to have less exception and less assumptions when handling tell/asok commands in our code. IMHO, the exceptions/assumptions are another form of magic numbers embedded in the code. if we have to live with the magic numbers, we need to make them explicit.

> Their signatures are passed register_command and the asok machinery checks them.

`AdminSocket::execute_command()` does check them. but it does not check cap or if the command is deprecated. i agree that it's far from ideal as we are doing almost the same thing in three different places: `Monitor::handle_tell_command()`, `Monitor::handle_command()` and `AdminSocket::execute_command()`. but i think it would be simpler if we repeat the logic in current way.

since we are able to direct `ceph` cli to a certain daemon to run any asok command, there is no reason to make the existing asok commands exceptions when it comes to cap check, and only apply the checks to a subset of them, in this case, the "smart" command belongs to this subset.

also, i think it'd be nice if we can have a process for deprecating user-facing commands. that's also why i added the declarations for the asok/tell commands -- so they can be deprecated in future.

Reply:
Hmm, I guess so.  My thinking was to only have the ~4 commands listed here as a compat kludge (that unfortunately would still around for a long time), but this works too.   If we do this, can we rename NOFORWARD to TELL?  And maybe make tell imply hidden so you don't have to specify both flags?

Reply:
updated accordingly in the latest changeset.